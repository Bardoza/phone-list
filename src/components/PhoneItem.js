import React, { Component } from "react";
import PropTypes from "prop-types";

class PhoneItem extends Component {
  static propTypes = {
    onChange: PropTypes.func.isRequired,
    index: PropTypes.number.isRequired,
    onDelete: PropTypes.func.isRequired,
    types: PropTypes.arrayOf(PropTypes.string).isRequired,
    availableTypes: PropTypes.arrayOf(PropTypes.string).isRequired,
    type: PropTypes.string.isRequired,
    number: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired
  };

  onTypeChange = e => {
    this.props.onChange(e.target.value, this.props.index, "type");
  };

  onNumberChange = e => {
    const number = e.target.value.replace(/\D/gi, "");
    if (number.length > 15) {
      return;
    }

    this.props.onChange(number, this.props.index, "number");
  };

  onDelete = () => this.props.onDelete(this.props.index);

  render() {
    return (
      <div className="phone-item">
        <div className="type-container">
          <select value={this.props.type} onChange={this.onTypeChange}>
            {this.props.types.map(type => (
              <option
                key={type}
                value={type}
                disabled={!this.props.availableTypes.some(t => t === type)}
              >
                {type}
              </option>
            ))}
          </select>
        </div>
        <input
          type="text"
          value={this.props.number}
          onChange={this.onNumberChange}
        />
        <button onClick={this.onDelete}>x</button>
      </div>
    );
  }
}

export default PhoneItem;
