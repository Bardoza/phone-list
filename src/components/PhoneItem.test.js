import React from 'react'
import { shallow } from 'enzyme'
import PhoneItem from './PhoneItem'

/**
 *  Proptypes
 *  onChange: PropTypes.func.isRequired,
    index: PropTypes.number.isRequired,
    onDelete: PropTypes.func.isRequired,
    types: PropTypes.arrayOf(PropTypes.string).isRequired,
    availableTypes: PropTypes.arrayOf(PropTypes.string).isRequired,
    type: PropTypes.string.isRequired,
    number: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired
 */

let phoneItem;
let mockOnChange;
let mockOnDelete; 

beforeEach( () => {
    mockOnChange = jest.fn()
    mockOnDelete = jest.fn()
    phoneItem = <PhoneItem 
    onChange={mockOnChange}
    onDelete={mockOnDelete}
    index={0}
    types={["Work", "Mobile"]}
    availableTypes={["Mobile"]}
    type="Work"
    number="476887251"
/>
})
describe('PhoneItem', () => {
    
    it("Render correctly", () => {
        expect(shallow(phoneItem)).toMatchSnapshot()  
    })

    it("should call onDelete when delete button is pressed", () => {
        const wrapper = shallow(phoneItem)
        wrapper.find('button').simulate("click")
        expect(mockOnDelete).toHaveBeenCalledTimes(1)
    })

    it("should render only 2 options", () => {
      
        const wrapper = shallow(phoneItem)
        expect(wrapper.find('option')).toHaveLength(2)
    })

    it("option with value Work should be disabled", () => {
        const wrapper = shallow(phoneItem)
        expect(
            wrapper.find('option[value="Work"]').prop("disabled")
        ).toBeTruthy()
    })
    it("option with value Mobile should be disabled", () => {
        const wrapper = shallow(phoneItem)
        expect(
            wrapper.find('option[value="Mobile"]').prop("disabled")
        ).toBeFalsy()
    })

    it("should call mockOnChange when changing select option", ()=> {
        const wrapper = shallow(phoneItem)
        wrapper.find('select').simulate('change',{target:{value:'Mobile'}})
        expect(mockOnChange).toBeCalledWith("Mobile", 0, "type")
    })

    it("should call mockOnChange when changing input text", ()=> {
        const wrapper = shallow(phoneItem)
        wrapper.find('input').simulate('change',{target:{value:'3548321'}})
        expect(mockOnChange).toBeCalledWith("3548321", 0, "number")
    })

    it("should call mockOnChange with only numerical values", ()=> {
        const wrapper = shallow(phoneItem)
        wrapper.find('input').simulate('change',{target:{value:'123456789hgsj'}})
        expect(mockOnChange).toBeCalledWith("123456789", 0, "number")
    })
    
})