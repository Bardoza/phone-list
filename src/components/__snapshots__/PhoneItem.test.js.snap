// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhoneItem Render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PhoneItem
    availableTypes={
        Array [
            "Mobile",
          ]
    }
    index={0}
    number="476887251"
    onChange={[Function]}
    onDelete={[Function]}
    type="Work"
    types={
        Array [
            "Work",
            "Mobile",
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="type-container"
>
          <select
                    onChange={[Function]}
                    value="Work"
          >
                    <option
                              disabled={true}
                              value="Work"
                    >
                              Work
                    </option>
                    <option
                              disabled={false}
                              value="Mobile"
                    >
                              Mobile
                    </option>
          </select>
</div>,
        <input
          onChange={[Function]}
          type="text"
          value="476887251"
/>,
        <button
          onClick={[Function]}
>
          x
</button>,
      ],
      "className": "phone-item",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <select
            onChange={[Function]}
            value="Work"
>
            <option
                        disabled={true}
                        value="Work"
            >
                        Work
            </option>
            <option
                        disabled={false}
                        value="Mobile"
            >
                        Mobile
            </option>
</select>,
          "className": "type-container",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <option
                disabled={true}
                value="Work"
>
                Work
</option>,
              <option
                disabled={false}
                value="Mobile"
>
                Mobile
</option>,
            ],
            "onChange": [Function],
            "value": "Work",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "Work",
              "nodeType": "host",
              "props": Object {
                "children": "Work",
                "disabled": true,
                "value": "Work",
              },
              "ref": null,
              "rendered": "Work",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "Mobile",
              "nodeType": "host",
              "props": Object {
                "children": "Mobile",
                "disabled": false,
                "value": "Mobile",
              },
              "ref": null,
              "rendered": "Mobile",
              "type": "option",
            },
          ],
          "type": "select",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "onChange": [Function],
          "type": "text",
          "value": "476887251",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "x",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "x",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="type-container"
>
            <select
                        onChange={[Function]}
                        value="Work"
            >
                        <option
                                    disabled={true}
                                    value="Work"
                        >
                                    Work
                        </option>
                        <option
                                    disabled={false}
                                    value="Mobile"
                        >
                                    Mobile
                        </option>
            </select>
</div>,
          <input
            onChange={[Function]}
            type="text"
            value="476887251"
/>,
          <button
            onClick={[Function]}
>
            x
</button>,
        ],
        "className": "phone-item",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <select
              onChange={[Function]}
              value="Work"
>
              <option
                            disabled={true}
                            value="Work"
              >
                            Work
              </option>
              <option
                            disabled={false}
                            value="Mobile"
              >
                            Mobile
              </option>
</select>,
            "className": "type-container",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <option
                  disabled={true}
                  value="Work"
>
                  Work
</option>,
                <option
                  disabled={false}
                  value="Mobile"
>
                  Mobile
</option>,
              ],
              "onChange": [Function],
              "value": "Work",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "Work",
                "nodeType": "host",
                "props": Object {
                  "children": "Work",
                  "disabled": true,
                  "value": "Work",
                },
                "ref": null,
                "rendered": "Work",
                "type": "option",
              },
              Object {
                "instance": null,
                "key": "Mobile",
                "nodeType": "host",
                "props": Object {
                  "children": "Mobile",
                  "disabled": false,
                  "value": "Mobile",
                },
                "ref": null,
                "rendered": "Mobile",
                "type": "option",
              },
            ],
            "type": "select",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onChange": [Function],
            "type": "text",
            "value": "476887251",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "x",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "x",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
